# CNN Basics

#### Name: Vishal L

#### Reg.No: 20BAI1038

### MNIST DATASET: 

In [1]:
# keras imports for the dataset and building our neural network
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Conv2D, MaxPool2D, Flatten
from keras.utils import np_utils

In [2]:
# to calculate accuracy
from sklearn.metrics import accuracy_score

In [3]:
# loading the dataset
(X_train, y_train), (X_test, y_test) = mnist.load_data()

11490434/11490434 [==============================] - 6s 1us/step


In [4]:
# building the input vector from the 28x28 pixels
X_train = X_train.reshape(X_train.shape[0], 28, 28, 1)
X_test = X_test.reshape(X_test.shape[0], 28, 28, 1)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

In [5]:
# normalizing the data to help with the training
X_train /= 255
X_test /= 255


In [6]:
# one-hot encoding using keras' numpy-related utilities
n_classes = 10
print("Shape before one-hot encoding: ", y_train.shape)
Y_train = np_utils.to_categorical(y_train, n_classes)
Y_test = np_utils.to_categorical(y_test, n_classes)
print("Shape after one-hot encoding: ", Y_train.shape)


Shape before one-hot encoding:  (60000,)
Shape after one-hot encoding:  (60000, 10)


In [7]:
# building a linear stack of layers with the sequential model
model = Sequential()

In [8]:
# convolutional layer
model.add(Conv2D(25, kernel_size=(3,3), strides=(1,1), padding='valid', activation='relu', input_shape=(28,28,1)))
model.add(MaxPool2D(pool_size=(1,1)))

In [9]:
# flatten output of conv
model.add(Flatten())

In [10]:
# hidden layer
model.add(Dense(100, activation='relu'))

In [11]:
# output layer
model.add(Dense(10, activation='softmax'))

In [12]:
# compiling the sequential model
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')

In [13]:
# training the model for 10 epochs
model.fit(X_train, Y_train, batch_size=128, epochs=10, validation_data=(X_test, Y_test))

Epoch 1/10
469/469 [==============================] - 25s 50ms/step - loss: 0.2010 - accuracy: 0.9426 - val_loss: 0.0786 - val_accuracy: 0.9763
Epoch 2/10
469/469 [==============================] - 23s 49ms/step - loss: 0.0622 - accuracy: 0.9811 - val_loss: 0.0608 - val_accuracy: 0.9798
Epoch 3/10
469/469 [==============================] - 23s 48ms/step - loss: 0.0391 - accuracy: 0.9880 - val_loss: 0.0537 - val_accuracy: 0.9816
Epoch 4/10
469/469 [==============================] - 23s 49ms/step - loss: 0.0253 - accuracy: 0.9923 - val_loss: 0.0485 - val_accuracy: 0.9835
Epoch 5/10
469/469 [==============================] - 22s 47ms/step - loss: 0.0170 - accuracy: 0.9951 - val_loss: 0.0482 - val_accuracy: 0.9853
Epoch 6/10
469/469 [==============================] - 23s 49ms/step - loss: 0.0100 - accuracy: 0.9973 - val_loss: 0.0525 - val_accuracy: 0.9849
Epoch 7/10
469/469 [==============================] - 23s 48ms/step - loss: 0.0067 - accuracy: 0.9982 - val_loss: 0.0571 - val_accuracy:

#### Even though our max validation accuracy by using a simple neural network model was around 97%, the CNN model is able to get 98%+ with just a single convolution layer!

## CIFAR-10 Dataset

In [14]:
# keras imports for the dataset and building our neural network
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Conv2D, MaxPool2D, Flatten
from keras.utils import np_utils

In [15]:
# loading the dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

170498071/170498071 [==============================] - 142s 1us/step


In [16]:
# # building the input vector from the 32x32 pixels
X_train = X_train.reshape(X_train.shape[0], 32, 32, 3)
X_test = X_test.reshape(X_test.shape[0], 32, 32, 3)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

In [17]:
# normalizing the data to help with the training
X_train /= 255
X_test /= 255

In [18]:
# one-hot encoding using keras' numpy-related utilities
n_classes = 10
print("Shape before one-hot encoding: ", y_train.shape)
Y_train = np_utils.to_categorical(y_train, n_classes)
Y_test = np_utils.to_categorical(y_test, n_classes)
print("Shape after one-hot encoding: ", Y_train.shape)

Shape before one-hot encoding:  (50000, 1)
Shape after one-hot encoding:  (50000, 10)


In [19]:
# building a linear stack of layers with the sequential model
model = Sequential()


In [20]:
# convolutional layer
model.add(Conv2D(50, kernel_size=(3,3), strides=(1,1), padding='same', activation='relu', input_shape=(32, 32, 3)))

In [21]:
# convolutional layer
model.add(Conv2D(75, kernel_size=(3,3), strides=(1,1), padding='same', activation='relu'))
model.add(MaxPool2D(pool_size=(2,2)))
model.add(Dropout(0.25))

In [22]:
model.add(Conv2D(125, kernel_size=(3,3), strides=(1,1), padding='same', activation='relu'))
model.add(MaxPool2D(pool_size=(2,2)))
model.add(Dropout(0.25))

In [23]:
# flatten output of conv
model.add(Flatten())


In [24]:
# hidden layer
model.add(Dense(500, activation='relu'))
model.add(Dropout(0.4))
model.add(Dense(250, activation='relu'))
model.add(Dropout(0.3))

In [25]:
# output layer
model.add(Dense(10, activation='softmax'))

In [26]:
# compiling the sequential model
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')

In [27]:
# training the model for 10 epochs
model.fit(X_train, Y_train, batch_size=128, epochs=10, validation_data=(X_test, Y_test))

Epoch 1/10
391/391 [==============================] - 163s 416ms/step - loss: 1.5843 - accuracy: 0.4176 - val_loss: 1.2420 - val_accuracy: 0.5613
Epoch 2/10
391/391 [==============================] - 176s 450ms/step - loss: 1.1313 - accuracy: 0.5980 - val_loss: 1.0447 - val_accuracy: 0.6382
Epoch 3/10
391/391 [==============================] - 181s 463ms/step - loss: 0.9331 - accuracy: 0.6707 - val_loss: 0.8222 - val_accuracy: 0.7190
Epoch 4/10
391/391 [==============================] - 171s 437ms/step - loss: 0.8154 - accuracy: 0.7165 - val_loss: 0.7589 - val_accuracy: 0.7362
Epoch 5/10
391/391 [==============================] - 171s 438ms/step - loss: 0.7308 - accuracy: 0.7447 - val_loss: 0.7011 - val_accuracy: 0.7560
Epoch 6/10
391/391 [==============================] - 177s 453ms/step - loss: 0.6604 - accuracy: 0.7702 - val_loss: 0.6769 - val_accuracy: 0.7655
Epoch 7/10
391/391 [==============================] - 169s 433ms/step - loss: 0.6079 - accuracy: 0.7871 - val_loss: 0.6601 -

## Arabic Handwritten digits Dataset

In [37]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import matplotlib.image as mpimg

from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import KFold

import tensorflow as tf
from keras.models import Sequential
from keras.layers import Dense, Conv2D, Flatten,MaxPool2D,Dropout,LSTM
from keras.optimizers import RMSprop, Adam, Nadam
from keras.callbacks import ReduceLROnPlateau
from keras.preprocessing.image import ImageDataGenerator
from keras.utils.np_utils import to_categorical

In [38]:
y_train = pd.read_csv('csvTrainLabel 60k x 1.csv')
X_train = pd.read_csv('csvTrainImages 60k x 784.csv')
y_test = pd.read_csv('csvTestLabel 10k x 1.csv')
X_test = pd.read_csv('csvTestImages 10k x 784.csv')

In [39]:
# Normalise and reshape
X_train=X_train.values.reshape((-1,28,28,1))/255.0
X_test=X_test.values.reshape((-1,28,28,1))/255.0

In [40]:
# OnehotEncode y_train
y_train=to_categorical(y_train, num_classes=10)
y_train

array([[0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]], dtype=float32)

In [41]:
# Split into Train and validation
X_train,X_valid,y_train, y_valid = train_test_split(X_train, y_train,
                                                     test_size=0.1,
                                                     shuffle=True)
# Data Augmentation to reduce bias
datagen = ImageDataGenerator(
        featurewise_center=False,  # set input mean to 0 over the dataset
        samplewise_center=False,  # set each sample mean to 0
        featurewise_std_normalization=False,  # divide inputs by std of the dataset
        samplewise_std_normalization=False,  # divide each input by its std
        zca_whitening=False,  # apply ZCA whitening
        #rotation_range=10,  # randomly rotate images in the range (degrees, 0 to 180)
        zoom_range = 0.1, # Randomly zoom image 
        width_shift_range=0.1,  # randomly shift images horizontally (fraction of total width)
        height_shift_range=0.1,  # randomly shift images vertically (fraction of total height)
        horizontal_flip=False,  # randomly flip images
        vertical_flip=False)  # randomly flip images


datagen.fit(X_train)

In [42]:
# Early Stopping
class custom_callback(tf.keras.callbacks.Callback):
    def on_epoch_end(self,epoch,logs={}):
        if(logs.get('accuracy')>0.92):
            self.model.stop_training=True
            
# LR Scheduler
learning_rate_reduction = ReduceLROnPlateau(monitor='val_acc', 
                                            patience=3, 
                                            verbose=1, 
                                            factor=0.5, 
                                            min_lr=0.00001)

In [43]:
# MODEL
def build_model():
    model = Sequential()
    model.add(Conv2D(filters=32, kernel_size=(5,5),padding='Same',
                activation='relu', input_shape=(28,28,1)))
    model.add(Conv2D(filters=32, kernel_size=(5,5),padding='Same',
                activation='relu', input_shape=(28,28,1)))
    model.add(MaxPool2D(pool_size=(2,2)))
    model.add(Dropout(0.3))

    model.add(Conv2D(filters=64, kernel_size=(3,3),padding='Same',
                activation='relu', input_shape=(28,28,1)))
    model.add(Conv2D(filters=64, kernel_size=(3,3),padding='Same',
                activation='relu', input_shape=(28,28,1)))
    model.add(MaxPool2D(pool_size=(2,2)))
    model.add(Dropout(0.4))

    model.add(Flatten())
    model.add(Dense(256,activation='relu'))
    model.add(Dropout((0.4)))
    model.add(Dense(10,activation='softmax'))
    return model

In [45]:
# Define the optimizer
optimizer = RMSprop(learning_rate=0.001, rho=0.9, epsilon=1e-08, decay=0.0)
# Compile the model
model = build_model()
model.compile(optimizer = optimizer , loss = "categorical_crossentropy", metrics=["accuracy"])

In [47]:
epochs=10
batch_size=64
# Fit the model
history = model.fit(datagen.flow(X_train,y_train, batch_size=batch_size),
                              epochs = epochs, validation_data = (X_valid,y_valid),
                              verbose = 2, callbacks=[learning_rate_reduction])

Epoch 1/10
844/844 - 128s - loss: 0.1783 - accuracy: 0.9451 - val_loss: 0.0351 - val_accuracy: 0.9898 - lr: 0.0010 - 128s/epoch - 151ms/step
Epoch 2/10
844/844 - 129s - loss: 0.0564 - accuracy: 0.9837 - val_loss: 0.0298 - val_accuracy: 0.9937 - lr: 0.0010 - 129s/epoch - 153ms/step
Epoch 3/10
844/844 - 129s - loss: 0.0491 - accuracy: 0.9868 - val_loss: 0.0258 - val_accuracy: 0.9943 - lr: 0.0010 - 129s/epoch - 152ms/step
Epoch 4/10
844/844 - 128s - loss: 0.0461 - accuracy: 0.9876 - val_loss: 0.0284 - val_accuracy: 0.9953 - lr: 0.0010 - 128s/epoch - 152ms/step
Epoch 5/10
844/844 - 129s - loss: 0.0467 - accuracy: 0.9880 - val_loss: 0.0221 - val_accuracy: 0.9952 - lr: 0.0010 - 129s/epoch - 152ms/step
Epoch 6/10
844/844 - 128s - loss: 0.0465 - accuracy: 0.9887 - val_loss: 0.0250 - val_accuracy: 0.9943 - lr: 0.0010 - 128s/epoch - 152ms/step
Epoch 7/10
844/844 - 132s - loss: 0.0460 - accuracy: 0.9887 - val_loss: 0.0255 - val_accuracy: 0.9945 - lr: 0.0010 - 132s/epoch - 156ms/step
Epoch 8/10
84

In [48]:
y_pred = model.predict(X_test)
y_test = to_categorical(y_test,num_classes=10)
model.evaluate(X_test,y_test)

313/313 [==============================] - 5s 17ms/step - loss: 0.0491 - accuracy: 0.9917


[0.049127764999866486, 0.9916991591453552]

### using Hyperas we increased the accuracy from 0.9451 to 0.9917